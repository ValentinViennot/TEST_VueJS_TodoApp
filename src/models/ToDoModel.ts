export default interface ToDoModel {
  text: string;
  done: boolean;
  id: number;
}
