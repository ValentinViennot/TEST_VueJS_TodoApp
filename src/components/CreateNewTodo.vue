<template>
  <div class="addContainer">
      <input class="addInput" type="text" v-model="newTodo.text" @keyup.enter="handleAdd" />
      <button class="addButton" @click="handleAdd">Add</button>
  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator';
import ToDoModel from '../models/ToDoModel';

@Component
export default class CreateNewTodo extends Vue {
  newTodo: ToDoModel = CreateNewTodo.getEmptyTodo();

  handleAdd() {
    if (this.isTodoEmpty()) return;
    this.$emit('new-todo', this.newTodo);
    this.newTodo = CreateNewTodo.getEmptyTodo();
  }

  private isTodoEmpty() {
    return this.newTodo.text.length === 0;
  }

  static getEmptyTodo() {
    return {
      text: '',
      done: false,
      id: -1
    };
  }
}
</script>

<style scoped>
</style>
