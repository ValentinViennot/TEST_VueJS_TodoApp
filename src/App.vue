<template>
  <div id="app">
    <h1>
      ToDo list 
    </h1>
    <div class="todosContainer">
      <ToDo v-for="todo of todos" v-bind:key="todo.id" v-bind:todo="todo" />
    </div>
    <CreateNewTodo v-on:new-todo="handleNewTodo" />
  </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';
import ToDo from './components/ToDo.vue';
import CreateNewTodo from './components/CreateNewTodo.vue';
import ToDoModel from '@/models/ToDoModel';

@Component({
  components: {
    ToDo,
    CreateNewTodo
  }
})
export default class App extends Vue {
  todos: ToDoModel[] = require('./assets/todos.json');

  handleNewTodo(newTodo: ToDoModel) {
    this.todos.push(newTodo);
  }
}
</script>

<style>
#app {
  font-family: 'Avenir', Helvetica, Arial, sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  text-align: center;
  color: #2c3e50;
  margin-top: 60px;
}

.todosContainer {
  width: 50%;
  min-width: 300px;
  margin: auto;
}
</style>
