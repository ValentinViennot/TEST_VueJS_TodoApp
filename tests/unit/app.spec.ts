import { shallowMount, Wrapper } from '@vue/test-utils';
import App from '../../src/App.vue';
import CreateNewTodo from '../../src/components/CreateNewTodo.vue';
import ToDoModel from '@/models/ToDoModel';

describe('App.vue component', () => {
  let wrapper: Wrapper<App>;

  beforeAll(() => {
    wrapper = shallowMount(App);
  });

  it('should render without crashing', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('when CreateNewTodo emit a new-todo event', () => {
    const dummyTodo: ToDoModel = {
      text: 'Hello',
      done: true,
      id: 4
    };

    beforeAll(() => {
      wrapper.find(CreateNewTodo).vm.$emit('new-todo', dummyTodo);
    });

    it('should push the emitted Todo to todos list', () => {
      expect(wrapper.vm.$data.todos.indexOf(dummyTodo)).not.toBe(-1);
    });
  });
});
