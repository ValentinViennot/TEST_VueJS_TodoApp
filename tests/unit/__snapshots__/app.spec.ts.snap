// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App.vue component should render without crashing 1`] = `
<div id="app">
  <h1>
    ToDo list
  </h1>
  <div class="todosContainer">
    <todo-stub todo="[object Object]"></todo-stub>
    <todo-stub todo="[object Object]"></todo-stub>
    <todo-stub todo="[object Object]"></todo-stub>
  </div>
  <createnewtodo-stub></createnewtodo-stub>
</div>
`;
